agent/utils/utils.go

```diff
@@ -109,11 +110,28 @@ var _time ttime.Time = &ttime.DefaultTime{}
 // If the error is Retriable then that will be used to determine if it should be
 // retried
 func RetryWithBackoff(backoff Backoff, fn func() error) error {
+	return RetryWithBackoffCtx(context.Background(), backoff, fn)
+}
+
+// RetryWithBackoffCtx takes a context, a Backoff, and a function to call that returns an error
+// If the context is done, nil will be returned
+// If the error is nil then the function will no longer be called
+// If the error is Retriable then that will be used to determine if it should be
+// retried
+func RetryWithBackoffCtx(ctx context.Context, backoff Backoff, fn func() error) error {
 	var err error
-	for err = fn(); true; err = fn() {
-		retriable, isRetriable := err.(Retriable)
+	for {
+		select {
+		case <-ctx.Done():
+			return nil
+		default:
+		}
 
-		if err == nil || isRetriable && !retriable.Retry() {
+		err = fn()
+
+		retriableErr, isRetriableErr := err.(Retriable)
+
+		if err == nil || isRetriableErr && !retriableErr.Retry() {
```

Comment:
Can you use parenthesis here to illustrate/enforce the operator precedence? Otherwise, readers like me will always have to look up the operator precedence documentation for how this is evaluated.

Reply:
It's just short-circuiting evaluation; any time a left-condition evaluates false, nothing proceeds.

Are you asking me to change this to `err == nil || (isRetriableErr && !retriableErr.Retry())`?  That's logically equivalent, but (to me) harder to read.

Reply:
> Are you asking me to change this to `err == nil || (isRetriableErr && !retriableErr.Retry())`?  That's logically equivalent, but (to me) harder to read.

Yeah, I prefer that. Because, then I wouldn't have to wonder if `isRetriableErr` is associated with the `||` or `&&` operator.